webui/react/src/hooks/useSettings.tsx

```diff
@@ -0,0 +1,332 @@
+    */
+    try {
+      const paramValue = params[setting.storageKey];
+      const baseType = setting.type.baseType;
+      const expectingArray = setting.type.isArray;
+
+      if (paramValue !== null) {
+        let queryValue: Primitive | Primitive[] | undefined;
+        /*
+        * Convert the string-based query params to primitives.
+        * `undefined` values can happen if the query param values are invalid.
+        *   string[] => Primitive[] | undefined
+        *   string   => Primitive | undefined
+        */
+        if (Array.isArray(paramValue)) {
+          queryValue = paramValue
+            .reduce<Primitive[]>((acc, value) => {
+              const parsedValue = queryParamToType(baseType.name as BaseType, value);
+
+              if (parsedValue !== undefined) acc.push(parsedValue);
+
+              return acc;
+            }, []);
+        } else {
+          queryValue = queryParamToType(baseType.name as BaseType, paramValue);
+        }
+
+        if (queryValue !== undefined) {
+          /*
+          * When expecting an array, convert valid non-array values into an array.
+          * Example - 'PULLING' => [ 'PULLING' ]
+          */
+          const normalizedValue = (() => {
+            if (expectingArray && !Array.isArray(queryValue)) {
+              return [queryValue];
+            }
+            return queryValue;
+          })();
+
+          acc[setting.storageKey] = normalizedValue;
+        }
+      }
+    } catch (e) {
+      handleError(e, { silent: true, type: ErrorType.Ui });
+    }
+
+    return acc;
+  }, {});
+};
+
+const useSettings = <T, >(config: SettingsConfig<T>): UseSettingsReturn => {
+  const { auth: { user } } = useStore();
+  const { state, update } = useContext(UserSettings);
+  const navigate = useNavigate();
+
+  // creates an entry at the global state
+  useEffect(() => {
+    const settings = state.get(config.applicableRoutespace);
+
+    if (settings) return;
+
+    const newSettings: Settings = {};
+
+    for (const key in config.settings) {
+      const setting = config.settings[key];
+
+      newSettings[setting.storageKey] = setting.defaultValue;
+    };
+
+    state.set(config.applicableRoutespace, newSettings);
+  }, [config, state]);
+
+  // parse navigation url to state
+  useEffect(() => {
+    const url = window.location.toString();
+
+    if (!url.includes('?')) return;
+
+    const settings = queryToSettings<T>(config, url);
+    const stateSettings = state.get(config.applicableRoutespace);
+
+    if (isEqual(settings, stateSettings)) return;
+
+    Object.keys(settings).forEach((setting) => {
+      state.set(config.applicableRoutespace, { [setting]: settings[setting] });
+    });
+  }, [config, state]);
+
+  const slicedState = state.get(config.applicableRoutespace);
+
+  const settings: Settings = useMemo(() => slicedState ?? {}, [slicedState]); // TODO: test if changing it to return only the state.get affects re-rendering
```

Comment:
I think you might actually be more worried _missed_ re-renders with your current setup. `useMemo` only restricts recalculations, it doesn't create new ones. Since the map isn't part of any state we need to introduce a manual rerender when we update it.